Add option to check if a particle is a primary particle and some code robustness amendments (LArG4ParticleFilter)

absolution1 opened the pull request:
Pull request adds a new feature to my old code and adds in some safety measures.  Some more work should probably be done but it's probably best to leave the changes I've listed at the bottom for another pull request.

Additions 
- Optinally check whether a particle is a primary particle

Under the hood changes
 - Check the all FCL vectors are the same length
 - Check that restricted-value fcl parameters do indeed only have restricted values
 - const-ify the interesting vectors to maintain their equal lengths
 - Swap the index argument for unsigned int
 - Check that the element requests on the vectors are valid
 - Pass the art::Ptr by const reference rather than const value

Stuff for a later pull request
 - Other stylistic changes
 - Enums should also be used for the restricted values
 - The restricted values should be changed (I expect this is a breaking change)

absolution1 committed:
Additions
 - Optinally check whether a particle is a primary particle

Under the hood changes
 - Check the all FCL vectors are the same length
 - Check that restricted-value fcl parameters do indeed only have restricted values
 - const-ify the interesting vectors to maintain their equal lengths
 - Swap the index argument for unsigned int
 - Check that the element requests on the vectors are valid
 - Pass the art::Ptr by const reference rather than const value

Other stylistic changes should be made.
Enums should also be used for the restricted values.
The restricted values should be changed.

FNALbuild commented:
A new Pull Request was created by  @absolution1 (Dom Brailsford) for develop.

It involves the following packages:

larevt


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//610/console

absolution1 commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8846&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds= 
 +build

FNALbuild commented:
-lariat_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4562&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
+argoneut 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4706&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
-lariat_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4563&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
+argoneut 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4707&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
-icarus_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/648&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
-sbnd_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5021&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
+dune 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9489&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
-dune_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9488&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
-uboone_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5144&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

FNALbuild commented:
-uboone_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5143&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8845&builds="

ghost commented:
I am not familiar with the code but CI tests look good so I'll approve this one.

ghost commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

knoepfel commented on larevt/Filters/LArG4ParticleFilter_module.cc:
```diff
@@ -27,57 +27,62 @@ namespace filt{
+      bool MaxMomentumCheck(const art::Ptr<simb::MCParticle> &particle, const unsigned int index);
+      bool StartInTPCCheck(const art::Ptr<simb::MCParticle> &particle, const unsigned int index);
+      bool StopInTPCCheck(const art::Ptr<simb::MCParticle> &particle, const unsigned int index);
+      bool TPCTrajLengthCheck(const art::Ptr<simb::MCParticle> &particle, const unsigned int index);
```
How many of the  member  functions can be `const`-qualified?  For example:

```diff
- bool IsInterestingParticle(const art::Ptr<simb::MCParticle> &particle, const unsigned int index);
+ bool IsInterestingParticle(const art::Ptr<simb::MCParticle> &particle, const unsigned int index) const;
```

absolution1 replied:
Good spot.  Literally all of the functions (except for `art::filter`) can be.  I actually thought I'd already done them.  Serves me right for believing that '5 years ago' me knew what I was doing.

Thanks for looking over the PR!

knoepfel commented on larevt/Filters/LArG4ParticleFilter_module.cc:
```diff
@@ -27,57 +27,62 @@ namespace filt{
-      std::vector<int> fStopInTPC;
-      std::vector<double> fParticleMinTPCLength;
-      bool fRequireAllInterestingParticles;
+      const std::string fLArG4ModuleLabel;
+      const std::vector<int> fInterestingPDGs;
+      const std::vector<int> fIsPrimary;
+      const std::vector<double> fParticleMinMomentum;
+      const std::vector<double> fParticleMaxMomentum;
+      const std::vector<int> fStartInTPC;
+      const std::vector<int> fStopInTPC;
+      const std::vector<double> fParticleMinTPCLength;
+      const bool fRequireAllInterestingParticles;
```
Bravo!

absolution1 replied:
Thanks very much!

knoepfel reviewed (commented):
One of the best PRs I've seen--thanks for submitting it!  Just some minor comments for your consideration regarding `const`-qualified member functions.  Nothing major, however.

absolution1 committed:
Const qualify all of the functions (except for the virtual filter function).
Move opening function brackets onto new lines for more consistent styling.  More tweaks are needed later.

FNALbuild commented:
Pull request #8 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//617/console

knoepfel commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8898&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds= 
 +build

FNALbuild commented:
-lariat_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4595&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
+argoneut 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4741&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
+argoneut 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4740&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
-lariat_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4594&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
-icarus_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/685&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
-sbnd_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5062&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
-dune_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9570&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
-dune_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9569&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
-uboone_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5178&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

FNALbuild commented:
-uboone_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5177&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8897&builds="

lgarren commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests.
